### ISSUE TITLE
[no-undef] Linter does not report error when using undefined variables.

### ISSUE THREAD

**Issue by gkjohnson** [typescript-eslint/typescript-eslint#662]
**Repro**

`package.json`
```JSON
{
  "name": "typescript-test",
  "scripts": {
    "lint": "eslint ./src/* --ext js --ext ts"
  },
  "devDependencies": {
    "@typescript-eslint/eslint-plugin": "^1.11.0",
    "@typescript-eslint/parser": "^1.11.0",
    "eslint": "^6.0.1",
    "typescript": "^3.5.2"
  },
  "eslintConfig": {
    "parser": "@typescript-eslint/parser",
    "plugins": [
      "@typescript-eslint"
    ],
    "rules": {
      "no-undef": 1
    }
  }
}
```

`src/test.d.ts` OR `src/test.ts`
```TS
export class TestClass {
    field: UndefinedClass;
}
```

**Expected Result**

A `no-undef` rule error is reported because `UndefinedClass` is being used but has not been defined.

**Actual Result**

No errors are reported.

**Additional Info**

The error is correctly reported when doing the equivalent in a js file.

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `1.11.0` |
| `@typescript-eslint/parser`        | `1.11.0` |
| `TypeScript`                       | `3.5.2` |
| `ESLint`                           | `6.0.1` |
| `node`                             | `12.4.0` |
| `npm`                              | `6.9.0` |

**Comment by bradzacher** [comment]
We suggest turning off `no-undef` for typescript projects.
It can only ever report errors that are handled automatically (and better) by the typescript compiler itself.

**Comment by gkjohnson** [comment]
> It can only ever report errors that are handled automatically (and better) by the typescript compiler itself.

I'm not sure I follow -- by "it" do you mean the typescript eslint plugin? To elaborate I expect that this feature would be implemented as a new rule (such as `@typescript-eslint/no-undef`) that reports the TS2304 error the way the compiler / VSCode does for files.

Thanks!

**Comment by bradzacher** [comment]
Our parser is only a parser. It leverages the typescript compiler to parse and generate an AST, and provide type information. However we don't currently have have access to the diagnostics typescript provides, nor can we access things like its scope analyser.

As it stands right now the rule maintains its own logic for analysing scope, which has no idea about typescript type scope.

If you're looking for a way to have typescript compiler error TS2304 reported, the best way is via running the typescript compiler..

**Comment by kaicataldo** [comment]
For what it’s worth, we turned no-under off and just let tsc report this at my last job and it was never an issue. Most popular editors are able to report errors reported by tsc, and we would run `tsc —noEmit` and ESLint during the linting step of our CI process.

**Comment by bradzacher** [comment]
That was the implication of the first comment.
Recommend just turning this rule off because the typescript compiler can report the problem better than a custom written rule ever could.

Considering this rule's functionality is completely covered by the typescript compiler, rewriting the rule to work properly is low on the list of priorities...

**Comment by gkjohnson** [comment]
Got it -- thanks for the clear explanation! I'll look into using the typescript compiler to catch these types of issues.

**Comment by bradzacher** [comment]
Merging into #1856
